### PR TITLE
Dynamic type refactor: step 1

### PR DESCRIPTION
**Pull request by zasdfgbnm**
I started to rewrite our dynamic typing  (a.k.a. `EvaluatorValue`), and this is the first step: implement a brand new `DynamicType`. In the followup PR, I will use `DynamicType` to replace `EvaluatorValue`.

There are two motivations in doing so:
1. I am currently uncomfortable with how `EvaluatorValue` is implemented today: in overloaded operators, it does a manual checking on the actual type, and dispatch differently. This manual checking makes it very hard to extend it to support new types, and it is also very easy to make mistakes. As we are adding more types in the future, we need a more scalable approach.
2. As we are supporting scalars of more and more fancy dtypes, it no longer makes sense to make `Scalar` a template class. Instead, we should de-template it, and use a dynamic typing system to store its values.

So I implemented this new `DynamicType` template class. `DynamicType<T1, T2, ...>` is an enhanced version of `std::variant<T1, T2, ...>`. It overloads almost all C++ operators, so that using `DynamicType` is just like using a dynamic type language like Python. Operator overloading is automatically handled by template meta-programming, so there is no more hand check and cast. By using template meta-programming, we are able to cover exactly the operators needed, no more, no less. For example, `DynamicType<int, double> + DynamicType<int, double>` is a valid operation, but `DynamicType<int*, double*> + DynamicType<int*, double*>` is compilation error, because pointers does not have addition operator with another pointer. Please see the note `[Design of DynamicType]` for more details.

Implementing `DynamicType` requires very heavy template meta-programming, which is infamous for its difficulty in debugging and reading. In order to make the maintenance easier, I created our own type traits library in `type_traits.h`. It contains lots of utilities making very in-depth use of C++'s type system. Each utility does a small single job, for example, `cartesian_product` takes the cartesian product of an arbitrary number of tuples, and `ForAllTypes` does something like "`for (using type : types)`". All these utilities are tested and documented separately. By combining these utilities, we will be able to do advanced tasks like, for example: given three lists of types `Ts1`, `Ts2`, `Ts3`, can we find a `T1` in `Ts1`, `T2` in `Ts2`, and `T3` in `Ts3`, such that `f(T1, T2)` is well-formed and return `T3`?

In my view, `DynamicType` is better in every aspect than `EvaluatorValue`: it is smarter, has better operator coverage, is better documented, and more thoroughly tested.

In the future, `EvaluatorValue` will be replaced by `DynamicType` completely,
```
using EvaluatorValue = DynamicType<double, int64_t, bool, ...>; // all dtypes we support
```
and `Scalar` will be a (non-template) class that uses `EvaluatorValue` to store its value.

I tested this PR locally with g++ 13.1.1 and clang++ 15.0.7, both works. I didn't test any older compiler versions.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
Clang-format is failing, but this time, I don't want to fix it (although it is just a single button click), because I think it makes the code worse-looking. So I am requesting a version bump: https://github.com/NVIDIA/Fuser/issues/481

File: csrc/dynamic_type.h [hunk 1]
```diff
@@ -9,14 +9,539 @@
...
+// doing Custom12{} + Custom12{} results in a compilation error. It is a
+// compilation error because we know at compile time that none of them are
+// defined:
+//  - operator+(CustomType, CustomType)
+//  - operator+(CustomType, CustomType2)
+//  - operator+(CustomType2, CustomType)
+//  - operator+(CustomType2, CustomType2)
+// So we decide decide to not create the operator+ for Custom12.
+//
+// Also, beside requiring operator+(T1, T2) to be defined for some T1 and T2 in
```
Comment (naoyam) [inline csrc/dynamic_type.h:66]: nit: besides

File: csrc/dynamic_type.h [hunk 2]
```diff
@@ -9,14 +9,539 @@
...
+//   using Custom12 = DynamicType<CustomType, CustomType2>;
+// Then the operator+ on Custom12 should not be defined at compile time, and
+// doing Custom12{} + Custom12{} results in a compilation error. It is a
+// compilation error because we know at compile time that none of them are
+// defined:
+//  - operator+(CustomType, CustomType)
+//  - operator+(CustomType, CustomType2)
+//  - operator+(CustomType2, CustomType)
+//  - operator+(CustomType2, CustomType2)
+// So we decide decide to not create the operator+ for Custom12.
```
Comment (naoyam) [inline csrc/dynamic_type.h:64]:
```suggestion
// So we decide to not create the operator+ for Custom12.
```

**Review by naoyam — APPROVED** [review]
Not 100% certain, but looks good as the tests are pretty thorough.

Didn't know C++17 expanded meta programming capabilities so much

File: csrc/dynamic_type.h (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/dynamic_type.h]: fixed

File: csrc/dynamic_type.h (hunk 2 above)
Comment (zasdfgbnm) [reply csrc/dynamic_type.h]: fixed

File: .lintrunner.toml [hunk 3]
```diff
@@ -62,7 +62,7 @@ init_command = [
     'python3',
     'tools/linter/adapters/pip_init.py',
     '--dry-run={{DRYRUN}}',
-    'clang-tidy==15.0.2.1',
+    'clang-tidy==14.0.6',
```
Comment (zasdfgbnm) [inline .lintrunner.toml:65]: clang-tidy 15.0.2 has issue linting this PR, clang-tidy 15.0.7 fixes that issue, but unfortunately PyPI does not have 15.0.7, so I change it to 14.0.6 instead.

Comment (zasdfgbnm) [reply .lintrunner.toml:65]: cc @jjsjann123

**Pull request merged**